lib/mayaUsd/ufe/UsdUndoRenameCommand.cpp

```diff
@@ -42,44 +42,49 @@ namespace ufe {
 
 UsdUndoRenameCommand::UsdUndoRenameCommand(const UsdSceneItem::Ptr& srcItem, const Ufe::PathComponent& newName)
     : Ufe::UndoableCommand()
+    , _ufeSrcItem(srcItem)
+    , _ufeDstItem(nullptr)
+    , _prim(srcItem->prim())
+    , _stage(_prim.GetStage())
+    , _newName(newName.string())
+    , _oldName(_prim.GetName().GetString())
 {
-    const UsdPrim& prim = srcItem->prim();
-    _stage = prim.GetStage();
-    _ufeSrcItem = srcItem;
-    _usdSrcPath = prim.GetPath();
-
-    // Every call to rename() (through execute(), undo() or redo()) removes
-    // a prim, which becomes expired.  Since USD UFE scene items contain a
-    // prim, we must recreate them after every call to rename.
-    _usdDstPath = prim.GetParent().GetPath().AppendChild(TfToken(newName.string()));
-
-    _layer = MayaUsdUtils::defPrimSpecLayer(prim);
-    if (!_layer) {
-        std::string err = TfStringPrintf("No prim found at %s", prim.GetPath().GetString().c_str());
-        throw std::runtime_error(err.c_str());
-    }
-
     // if the current layer doesn't have any opinions that affects selected prim
-    if (!MayaUsdUtils::doesEditTargetLayerHavePrimSpec(prim)) {
-        auto possibleTargetLayer = MayaUsdUtils::strongestLayerWithPrimSpec(prim);
+    if (!MayaUsdUtils::doesEditTargetLayerHavePrimSpec(_prim)) {
+        auto possibleTargetLayer = MayaUsdUtils::strongestLayerWithPrimSpec(_prim);
         std::string err = TfStringPrintf("Cannot rename [%s] defined on another layer. " 
                                          "Please set [%s] as the target layer to proceed", 
-                                         prim.GetName().GetString().c_str(),
+                                         _prim.GetName().GetString().c_str(),
                                          possibleTargetLayer->GetDisplayName().c_str());
         throw std::runtime_error(err.c_str());
     }
     else
     {
-        auto layers = MayaUsdUtils::layersWithPrimSpec(prim);
+        // account for internal vs external references
+        if (_prim.HasAuthoredReferences()) {
+            auto primSpec = MayaUsdUtils::getPrimSpecAtEditTarget(_stage, _prim);
+            for (const SdfReference& ref : primSpec->GetReferenceList().GetAddedOrExplicitItems()) {
```

Comment:
Feels strange that this is going through a loop...  Because in the loop body, the if part of the conditional returns, and the else part of the conditional throws, so there is no way you'll get anything beyond the first item in the loop!

Also, is it possible for _prim.HasAuthoredReferences() to be true, but primSpec->GetReferenceList().GetAddedOrExplicitItmes() to return an empty list?

Reply:
@ppt-adsk  yes that was my first impression about the loop but that's what we have got. 

`Because in the loop body, the if part of the conditional returns, and the else part of the conditional throws, so there is no way you'll get anything beyond the first item in the loop!`

Yes I agree. This is a first pass and I am sure I am going to modify this code as I understand more edge cases.

`Also, is it possible for _prim.HasAuthoredReferences() to be true, but primSpec->GetReferenceList().GetAddedOrExplicitItmes() to return an empty list?`

I don't think that would be the case at least by looking at the low-level code. I could be wrong.